flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java

```diff
@@ -59,19 +73,30 @@
      * <p>NOTE: the action must be idempotent because multiple instances of it can be executed
      * concurrently (if the policy allows retries).
      */
-    void execute(RetryPolicy retryPolicy, RetriableAction action) {
+    void execute(
+            RetryPolicy retryPolicy, RetriableAction action, Consumer<Throwable> failureCallback) {
         LOG.debug("execute with retryPolicy: {}", retryPolicy);
         RetriableTask task =
-                new RetriableTask(action, retryPolicy, scheduler, attemptsPerTaskHistogram);
-        scheduler.submit(task);
+                RetriableTask.initialize(
+                        action,
+                        retryPolicy,
+                        blockingExecutor,
+                        attemptsPerTaskHistogram,
+                        timer,
+                        failureCallback);
+        blockingExecutor.submit(task);
     }
 
     @Override
     public void close() throws Exception {
         LOG.debug("close");
-        scheduler.shutdownNow();
-        if (!scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
-            LOG.warn("Unable to cleanly shutdown executorService in 1s");
+        timer.shutdownNow();
+        if (!timer.awaitTermination(1, TimeUnit.SECONDS)) {
```

Comment:
If `awaitTermination` throw exception out, the `blockingExecutor` would not shutdown.